File: 0654. Maximum Binary Tree/Solution.cpp
#include <algorithm>
#include <stack>
#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode()
        : val(0)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x)
        : val(x)
        , left(nullptr)
        , right(nullptr)
    {
    }
    TreeNode(int x, TreeNode* left, TreeNode* right)
        : val(x)
        , left(left)
        , right(right)
    {
    }
};

class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums)
    {
        vector<TreeNode*> stk;
        for (const int& e : nums) {
            TreeNode* p = new TreeNode(e);
            while (!stk.empty() && stk.back()->val < p->val) {
                p->left = stk.back();
                stk.pop_back();
            }
            if (!stk.empty())
                stk.back()->right = p;
            stk.emplace_back(p);
        }
        return stk.front();
    }
};

File: 0654. Maximum Binary Tree/Solution.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> Optional[TreeNode]:
        stack = []
        for e in nums:
            p = TreeNode(e)
            while stack and stack[-1].val < e:
                p.left = stack.pop()
            if stack:
                stack[-1].right = p
            stack.append(p)
        return stack[0]
